In [7]:
import sqlite3
import pandas as pd

In [8]:
csv_file_path = '/content/HR_Data.csv'
db_path = '/content/hr_data.db'

In [9]:
hr_data = pd.read_csv(csv_file_path)

In [10]:
conn = sqlite3.connect(db_path)

In [11]:
hr_data.to_sql('HRData', conn, if_exists='replace', index=False)

1470

# **How many employees are there in total, and how many of them have left the company (attrition)?**


In [18]:
query_employee_count_by_attrition = """
SELECT Attrition, COUNT(*) AS EmployeeCount
FROM HRData
GROUP BY Attrition;
"""
# Execute the SQL query and fetch the result into a DataFrame
df = pd.read_sql_query(query_employee_count_by_attrition, conn)

# Display the results
print(df)


  Attrition  EmployeeCount
0        No           1233
1       Yes            237


The total number of employees in the company is 1,470. Out of these, 237 employees have left the company, indicating an attrition rate of approximately 16.1%.

# **What is the average age of employees in each department**


In [19]:
query_average_age_by_department = """
SELECT Department, AVG(Age) AS AverageAge
FROM HRData
GROUP BY Department;
"""

# Execute the SQL query and fetch the result into a DataFrame
df = pd.read_sql_query(query_average_age_by_department, conn)

# Display the results
print(df)


               Department  AverageAge
0         Human Resources   37.809524
1  Research & Development   37.042664
2                   Sales   36.542601


The average age of employees varies slightly by department:

*   Human Resources: 37.8 years
*   Research & Development: 37.0 years
*   Sales: 36.5 years

Overall, the average age is fairly consistent across departments, with Human Resources having the highest average age and Sales the lowest.

#**What is the total monthly income generated by employees in each job role?**

In [20]:
query_total_monthly_income = """
        SELECT JobRole, SUM(MonthlyIncome) AS TotalMonthlyIncome
        FROM HRData
        GROUP BY JobRole;
"""

# Execute the SQL query and fetch the result into a DataFrame
df = pd.read_sql_query(query_total_monthly_income, conn)

# Display the results
print(df)

                     JobRole  TotalMonthlyIncome
0  Healthcare Representative              986268
1            Human Resources              220259
2      Laboratory Technician              838427
3                    Manager             1752531
4     Manufacturing Director             1057795
5          Research Director             1282684
6         Research Scientist              946072
7            Sales Executive             2257315
8       Sales Representative              217958


The total monthly income generated by employees varies significantly by job role: Sales Executive roles contribute the highest to total monthly income, while Human Resources and Sales Representative roles contribute the least.

# **How many employees who work overtime are satisfied with their jobs?**

In [23]:
query_overTime_job_satisfaction = """
        SELECT JobSatisfaction, COUNT(*) AS EmployeeCount
        FROM HRData
        WHERE OverTime = 'Yes'
        GROUP BY JobSatisfaction;
"""

# Execute the SQL query and fetch the result into a DataFrame
df = pd.read_sql_query(query_overTime_job_satisfaction, conn)

# Display the results
print(df)

   JobSatisfaction  EmployeeCount
0                1             84
1                2             69
2                3            121
3                4            142


Among employees who work overtime, job satisfaction levels are distributed as follows:



*   Satisfaction Level 1: 84 employees
*   Satisfaction Level 2: 69 employees

*   Satisfaction Level 3: 121 employees
*   Satisfaction Level 4: 142 employees

This indicates a range of satisfaction levels among those who work overtime, with the highest number of employees reporting the highest satisfaction level (4), and the lowest number reporting the lowest satisfaction level (1).

# **What is the average daily rate for male and female employees?**

In [24]:
query_avg_daily_rate_gender  = """
        SELECT Gender, AVG(DailyRate) AS AverageDailyRate
        FROM HRData
        GROUP BY Gender;
"""

# Execute the SQL query and fetch the result into a DataFrame
df = pd.read_sql_query(query_avg_daily_rate_gender, conn)

# Display the results
print(df)

   Gender  AverageDailyRate
0  Female        808.273810
1    Male        798.626984


This shows that, on average, female employees have a slightly higher daily rate compared to their male counterparts.

# **Who are the top 10 employees with the highest monthly income?**

In [25]:
query_top10_highest_income = """
SELECT *
FROM HRData
ORDER BY MonthlyIncome DESC
LIMIT 10;
"""

# Execute the SQL query and fetch the result into a DataFrame
df = pd.read_sql_query(query_top10_highest_income, conn)

# Display the results
print(df)

   Age Attrition     BusinessTravel  DailyRate              Department  \
0   52        No      Travel_Rarely        699  Research & Development   
1   41        No         Non-Travel        247  Research & Development   
2   56        No      Travel_Rarely        718  Research & Development   
3   50        No      Travel_Rarely       1452  Research & Development   
4   55       Yes      Travel_Rarely        725  Research & Development   
5   51        No  Travel_Frequently        237                   Sales   
6   52       Yes      Travel_Rarely        266                   Sales   
7   40        No      Travel_Rarely        611                   Sales   
8   43        No      Travel_Rarely        920  Research & Development   
9   56        No      Travel_Rarely        206         Human Resources   

   DistanceFromHome  Education    EducationField  EmployeeCount  \
0                 1          4     Life Sciences              1   
1                 7          1     Life Sciences   

The top 10 employees with the highest monthly income include a variety of job roles and departments. The highest earners are distributed among the Research & Development department, Sales, and Human Resources. Notably, these top earners vary in age, distance from home, and other attributes, reflecting diverse profiles among the highest income earners.

# **How many employees are there in each education field?**

In [26]:
query_employee_count_by_education = """
SELECT EducationField, COUNT(*) AS EmployeeCount
FROM HRData
GROUP BY EducationField;
"""
# Execute the SQL query and fetch the result into a DataFrame
df = pd.read_sql_query(query_employee_count_by_education, conn)

# Display the results
print(df)

     EducationField  EmployeeCount
0   Human Resources             27
1     Life Sciences            606
2         Marketing            159
3           Medical            464
4             Other             82
5  Technical Degree            132


The distribution of employees across different education fields shows that Life Sciences has the highest number of employees, with 606 individuals. Medical and Marketing fields follow with 464 and 159 employees, respectively. Human Resources and Technical Degree have fewer employees, at 27 and 132, while Other is the least represented with 82 employees.

# **What is the average number of years that employees who left the company worked there?**

In [27]:
query_avg_years_attrited = """
SELECT AVG(YearsAtCompany) AS AverageYears
FROM HRData
WHERE Attrition = 'Yes';
"""
df = pd.read_sql_query(query_avg_years_attrited, conn)

print(df)

   AverageYears
0      5.130802


The average number of years that employees who have left the company worked there is approximately 5.13 years. This indicates that, on average, employees who left the company had been with the company for just over five years before their departure.

# **What percentage of employees travel frequently for work?**


In [28]:
query_freq_travel_pct = """
SELECT (COUNT(*) * 100.0 / (SELECT COUNT(*) FROM HRData)) AS FrequentTravelPercentage
FROM HRData
WHERE BusinessTravel = 'Travel_Frequently';
"""

df = pd.read_sql_query(query_freq_travel_pct, conn)

print(df)

   FrequentTravelPercentage
0                 18.843537


Approximately 18.84% of employees travel frequently for work. This indicates that nearly one-fifth of the workforce is involved in frequent travel as part of their job responsibilities.









#  **What is the maximum number of years since the last promotion for each job level?**

In [29]:
query_max_years_promotion = """
SELECT JobLevel, MAX(YearsSinceLastPromotion) AS MaxYears
FROM HRData
GROUP BY JobLevel;
"""
df = pd.read_sql_query(query_max_years_promotion, conn)

print(df)

   JobLevel  MaxYears
0         1        15
1         2        13
2         3        15
3         4        15
4         5        15


For all job levels, the maximum number of years since the last promotion is 15 years. This suggests that regardless of job level, some employees have not received a promotion for up to 15 years, which could highlight a potential area for career development and advancement opportunities within the company.

# **What is the average job satisfaction level in each department?**


In [30]:
query_avg_job_satisfaction_dept = """
SELECT Department, AVG(JobSatisfaction) AS AverageJobSatisfaction
FROM HRData
GROUP BY Department;
"""
df = pd.read_sql_query(query_avg_job_satisfaction_dept, conn)

print(df)

               Department  AverageJobSatisfaction
0         Human Resources                2.603175
1  Research & Development                2.726327
2                   Sales                2.751121


The average job satisfaction levels vary slightly across departments, with the Sales department having the highest average satisfaction (2.75), followed by Research & Development (2.73), and Human Resources (2.60). This suggests that employees in Sales are generally more satisfied with their jobs compared to those in other departments, which could be a focal point for improving satisfaction in other areas.

# **How does marital status affect employee attrition rates?**

In [31]:
query_count_marital_status_attrition = """
SELECT MaritalStatus, Attrition, COUNT(*) AS EmployeeCount
FROM HRData
GROUP BY MaritalStatus, Attrition;
"""
df = pd.read_sql_query(query_count_marital_status_attrition, conn)

print(df)

  MaritalStatus Attrition  EmployeeCount
0      Divorced        No            294
1      Divorced       Yes             33
2       Married        No            589
3       Married       Yes             84
4        Single        No            350
5        Single       Yes            120


The data shows that marital status does influence employee attrition rates:



*   Divorced: 33 out of 327 (10.1%) employees left the company.
*   Married: 84 out of 673 (12.5%) employees left the company.

*   Single: 120 out of 470 (25.5%) employees left the company.





Single employees have the highest attrition rate, with about 25.5% leaving the company, compared to 12.5% for married employees and 10.1% for divorced employees. This indicates that single employees are more likely to leave the company than their married or divorced counterparts.

#  **What is the average distance from home for employees who have left the company?**


In [32]:
query_avg_distance_attrited = """
SELECT AVG(DistanceFromHome) AS AverageDistance
FROM HRData
WHERE Attrition = 'Yes';
"""
df = pd.read_sql_query(query_avg_distance_attrited, conn)

print(df)

   AverageDistance
0        10.632911


The average distance from home for employees who have left the company is approximately 10.63 miles. This suggests that employees who leave the company tend to live around this average distance from their workplace.

#  **What is the distribution of employees across different job levels?**


In [33]:
query_job_level_distribution = """
SELECT JobLevel, COUNT(*) AS EmployeeCount
FROM HRData
GROUP BY JobLevel
ORDER BY JobLevel;
"""
df = pd.read_sql_query(query_job_level_distribution, conn)

print(df)

   JobLevel  EmployeeCount
0         1            543
1         2            534
2         3            218
3         4            106
4         5             69


This distribution shows that the majority of employees are in the lower job levels, with fewer employees at higher levels.

#  **How do performance ratings vary across different job roles?**


In [34]:
query_avg_performance_by_role = """
SELECT JobRole, AVG(PerformanceRating) AS AveragePerformanceRating
FROM HRData
GROUP BY JobRole;
"""
df = pd.read_sql_query(query_avg_performance_by_role, conn)

print(df)

                     JobRole  AveragePerformanceRating
0  Healthcare Representative                  3.152672
1            Human Resources                  3.134615
2      Laboratory Technician                  3.162162
3                    Manager                  3.196078
4     Manufacturing Director                  3.186207
5          Research Director                  3.100000
6         Research Scientist                  3.167808
7            Sales Executive                  3.125767
8       Sales Representative                  3.144578


Overall, Managers and Manufacturing Directors have the highest average performance ratings, while Research Directors have the lowest. Most roles show similar performance ratings, with slight variations.

# **How many employees have stock options, and what are the levels of those options?**

In [35]:
query_stock_options_count = """
SELECT StockOptionLevel, COUNT(*) AS EmployeeCount
FROM HRData
GROUP BY StockOptionLevel;
"""
df = pd.read_sql_query(query_stock_options_count, conn)

print(df)

   StockOptionLevel  EmployeeCount
0                 0            631
1                 1            596
2                 2            158
3                 3             85


The majority of employees either have no stock options or are at Level 1. Fewer employees have higher levels of stock options.

# **How does job involvement relate to job satisfaction levels?**

In [36]:
query_job_involvement_satisfaction = """
SELECT JobInvolvement, AVG(JobSatisfaction) AS AverageJobSatisfaction
FROM HRData
GROUP BY JobInvolvement;
"""
df = pd.read_sql_query(query_job_involvement_satisfaction, conn)

print(df)

   JobInvolvement  AverageJobSatisfaction
0               1                2.915663
1               2                2.680000
2               3                2.745392
3               4                2.645833


The relationship between job involvement and job satisfaction shows that employees with higher job involvement tend to have lower average job satisfaction:This indicates a trend where employees with higher job involvement experience slightly lower job satisfaction.

# **What is the turnover rate for employees based on their business travel frequency?**


In [37]:
query_turnover_by_travel = """
SELECT BusinessTravel, (SUM(CASE WHEN Attrition = 'Yes' THEN 1 ELSE 0 END) * 100.0 / COUNT(*)) AS TurnoverRate
FROM HRData
GROUP BY BusinessTravel;
"""
df = pd.read_sql_query(query_turnover_by_travel, conn)

print(df)

      BusinessTravel  TurnoverRate
0         Non-Travel      8.000000
1  Travel_Frequently     24.909747
2      Travel_Rarely     14.956855


The turnover rate varies significantly based on business travel frequency:



*   Non-Travel: Turnover Rate = 8.00%
*   Travel Frequently: Turnover Rate = 24.91%
*   Travel Rarely: Turnover Rate = 14.96%


Employees who travel frequently for business have the highest turnover rate, suggesting that frequent travel may contribute to higher attrition.

# **How many training sessions did employees receive last year on average, broken down by job role?**


In [38]:
query_avg_training_by_role = """
SELECT JobRole, AVG(TrainingTimesLastYear) AS AverageTraining
FROM HRData
GROUP BY JobRole;
"""
df = pd.read_sql_query(query_avg_training_by_role, conn)

print(df)

                     JobRole  AverageTraining
0  Healthcare Representative         2.748092
1            Human Resources         2.557692
2      Laboratory Technician         2.949807
3                    Manager         2.813725
4     Manufacturing Director         2.744828
5          Research Director         2.775000
6         Research Scientist         2.664384
7            Sales Executive         2.831288
8       Sales Representative         3.012048


The average number of training sessions received last year varies by job role:

Sales Representatives received the highest average number of training sessions, while Research Scientists received the fewest.

# **How does the number of years an employee has been with their current manager relate to attrition?**

In [39]:
query_years_with_manager_attrition = """
SELECT YearsWithCurrManager, COUNT(*) AS EmployeeCount
FROM HRData
WHERE Attrition = 'Yes'
GROUP BY YearsWithCurrManager;
"""
df = pd.read_sql_query(query_years_with_manager_attrition, conn)

print(df)

    YearsWithCurrManager  EmployeeCount
0                      0             85
1                      1             11
2                      2             50
3                      3             19
4                      4             11
5                      5              4
6                      6              4
7                      7             31
8                      8             10
9                      9              6
10                    10              3
11                    11              1
12                    14              2


The number of years an employee has been with their current manager has a noticeable effect on attrition:



*   Employees with fewer years (0-1) with their current manager have a higher count of attrition cases.

*   Attrition decreases as the number of years with the current manager increases, with fewer employees leaving after more extended periods (10+ years).


This suggests that shorter tenures with a manager are associated with higher attrition rates.